Add `services.conf` for Simple API configuration.

leftclickben opened the pull request:
- Add object search and detail endpoints for library, museum,
  photographs and public memorials.
- Restricted to publicly accessible objects.
- Need to confirm the actual fields required to be displayed.
- No search for other types, need to check if this is necessary.

Commit:
Add `services.conf` for Simple API configuration.

+ Add object search and detail endpoints for library, museum,
  photographs and public memorials.
+ Restricted to publicly accessible objects.
+ Need to confirm the actual fields required to be displayed.
+ No search for other types, need to check if this is necessary.

leftclickben commented:
Not for merge yet, still need to check the fields from the spec.

kehh commented:
This actually looks pretty good.

kehh commented:
assigning to you while you're still working on it.

Commit:
Update based on checking specification.

+ No images included yet, that will be a separate PR.

leftclickben commented:
This is ready to go.  It doesn't include images, I'll do that separately.

Commit:
Add missing (and group) publication details.

Commit:
Add missing `Importance` field.

Commit:
Add missing `type_id` field.

kehh commented on conf/services.conf:
```diff
@@ -0,0 +1,163 @@
+            Publisher = ^ca_objects.Publisher,
+            DateOfPublication = ^ca_objects.DateOfPublication,
+            PlaceOfPublication = ^ca_objects.PlaceOfPublication,
+            PublicationType = ^ca_objects.PublicationType,
```
You may need to get it to get the list item label for these list attribute values, but I'm guessing that will come out in the wash.

leftclickben replied:
Pretty sure it comes with the actual value not the id so it's okay. The multivalued fields come out as semicolon separated lists of names which I can process client side. Creator is an interesting one.

leftclickben replied:
Confirmed this comes over as the label.

leftclickben commented on conf/services.conf:
```diff
@@ -0,0 +1,163 @@
+        content = {
+            object_id = ^ca_objects.object_id,
+            idno = ^ca_objects.idno,
+            type = ^ca_objects.type_id,
```
For some reason this doesn't seem to work, have tried a few variations, any ideas?

kehh replied:
`^type`? or `^type_id`

leftclickben replied:
Yep `^type_id` did it.

kehh commented on conf/services.conf:
```diff
@@ -0,0 +1,163 @@
+            DateOfPublication = ^ca_objects.DateOfPublication,
+            PlaceOfPublication = ^ca_objects.PlaceOfPublication,
+            PublicationType = ^ca_objects.PublicationType,
+            Subjects = ^ca_occurrences
```
According to the spec these should also be searchable (I see you've got them in the detail view):

```
            Series = ^ca_objects.Series,
            LibraryNumber = ^ca_objects.LibraryNumber,
            Donated = ^ca_objects.Donated
```

I'm also unclear as to whether Purchased, OrderNumber and Review should be added, but those can be added later.

leftclickben replied:
These aren't the fields available to search on, these are the columns available for the results table. Query  can be anything.

leftclickben replied:
Looks like I missed Purchased in library_detail though.

kehh commented on conf/services.conf:
```diff
@@ -0,0 +1,163 @@
+            ItemDates = ^ca_objects.ItemDates,
+            Importance = ^ca_objects.Importance,
+            Subjects = ^ca_occurrences,
+            Classification = ^ca_objects.Classification
```
Add:

```
             Description = ^ca_objects.Description,
             MakersMarks = ^ca_objects.MakersMarks,
             HistoricalDetails = ^ca_objects.HistoricalDetails,
             BibliographReferences = ^ca_objects.BibliographReferences,
             StatementOfSignificance = ^ca_objects.StatementOfSignificance
```

While these are text fields it doesn't hurt to search by them. Also wondering how we do `_fulltext`, but that can come later.

kehh replied:
Just realising that this isn't the fields that we want them to search by, it's the fields that are displayed after performing a search - am I right?

kehh replied:
If so please disregard pretty much all of the above comments.

leftclickben replied:
As per this previous  comment, yes :-)

https://github.com/rwahs/providence-configuration/pull/20/files/7d4feaa4f4aa186c3592ed30e154c3853c6bc563#r65115862

kehh commented on conf/services.conf:
```diff
@@ -0,0 +1,163 @@
+            Importance = ^ca_objects.Importance,
+            Subjects = ^ca_occurrences,
+            StatementOfSignificance = ^ca_objects.StatementOfSignificance,
+            Classification = ^ca_objects.Classification
```
There is no `Classification` attribute, this is done via a relationship to the controlled vocabulary. I _think_ this should be entered like:

`Classification = ^ca_list_items%restrictToRelationshipTypes=described`
but it might be:

```
Classification = <unit relativeTo="ca_list_items" restrictToRelationshipTypes="described">^ca_list_items.preferred_labels.name_singular</unit>
```

kehh commented on conf/services.conf:
```diff
@@ -0,0 +1,163 @@
+            HistoricalDetails = ^ca_objects.HistoricalDetails,
+            BibliographReferences = ^ca_objects.BibliographReferences,
+            Importance = ^ca_objects.Importance,
+            Subjects = ^ca_occurrences,
```
Subjects should include a restrictToTypes=subject (either using %restrictToTypes or a `<unit/>` tag.

kehh replied:
correct value is:

```
Subjects = ^ca_occurrences%restrictToTypes=subject
```

kehh commented:
Classification is done like this:

```
Classification = ^ca_list_items.hierarchy.preferred_labels.name_singular
```

Commit:
Change from review feedback.

+ Add `%restrictToTypes=subject` for subject relationships.
+ Add `%restrictToRelationshipTypes=described` for classifications.
+ Use correct template `^type_id` for type.

leftclickben commented:
All comments addressed, should be good to go now.

Commit:
Display hierarchy in Classification

Commit:
Configure Classification hierarchy output correctly.

Commit:
Changes to service configuration from review feedback.

+ Add missing `EarliestYear` and `LatestYear` fields to museum detail.
+ Add missing `Purchased` field to library detail.
+ Use templates to generate JSON strings for multi-value and container fields.
+ This unfortunately means that `removeFirstItems` no longer works, so the
  hierarchy root nodes are coming through and will need removal in the frontend.

leftclickben commented:
OK I think that's all... 4 R3@L$ this time.
